platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
+platform. This value should not change during address book changes and its value does not change unless the network
+goes through a another TSS bootstrap process. Since its value is not expected to change, storing it in a singleton
+Merkle Leaf by itself is appropriate.
+
+TODO: Protobuf of data
+
+##### Consensus Rosters
+
+This proposal extends the roster data format where each `RosterEntry` has a new long term public EC key called a
+`tssEcKey` that is used in the Groth21 algorithm.
+
+TODO: Protobuf of data
+
+##### Key Material (TSS Data Map)
+
+The TSS Data Map is a Map Merkle Leaf that is a combined map of all the key material for all rosters that are
+tracked and the votes by nodes that have validated the key material.
+
+Keys in the TSS Data Map have the following structure: (KeyType, RosterHash, SequenceNumber).
+
+1. KeyType - An enum that is one of TSS_MESSAGE or TSS_VOTE.
+2. RosterHash - The hash of the consensus roster that the message or vote is related to.
+3. SequenceNumber - The order in which the message or vote came to consensus.
+
+The value associated with a TSS_MESSAGE key is the pair (ShareId, TssMessage)
+
+1. ShareId - The id of the share that the message is related to.
+2. TssMessage - the raw bytes of the TssMessage
+
+The value associated with a TSS_VOTE key is a bit vector with the following interpretation
+
+* The order of bits from the least significant bit to the most significant bit corresponds to the numeric order of
+  share ids from least to greatest.
+* If a bit is set to 1, then the TssMessage for the corresponding ShareId was valid and contributed to a successful
+  reconstruction of the ledger id.
+* If a bit is set to 0, then the TssMessage for the corresponding ShareId was either invalid, was not received, or
+  was not used in the reconstruction of the ledger id.
+* If the sum of the bits is greater than the threshold, then the vote is a yes vote in favor of adopting the
+  candidate roster.
+
+Lifecycle Invariants
+
+1. Inserts should only be made for candidate rosters that are going through the keying process.
+2. When a roster is no longer stored in the state, the corresponding key material should be removed.
+3. Every roster that has become active after the TSS bootstrap process must have sufficient key material to be
+   able to recover the ledger id.
```

Comment:
> able to recover the ledger id

While this is correct, I'm not sure we need to bring up "recovering" the ledger ID. Really, what we are looking for is a threshold number of valid TSS messages. The fact that a threshold number of valid TSS messages can be used to recompute the ledger ID seems to me like merely a side effect.

Reply:
or a "double check your work" and throw an exception if this invariant is ever violated.